File: SDE_sheet/Arrays/11_missing_and_repeating_number/solution.go
func findMissingRepeatingNumbers(a []int) [2]int {
    n := len(a)
    xr := 0
    for i := 0; i < n; i++ {
        xr ^= a[i]
        xr ^= i + 1
    }
    number := xr & ^(xr - 1)
    zero, one := 0, 0
    for i := 0; i < n; i++ {
        if a[i]&number != 0 {
            one ^= a[i]
        } else {
            zero ^= a[i]
        }
    }
    for i := 1; i <= n; i++ {
        if i&number != 0 {
            one ^= i
        } else {
            zero ^= i
        }
    }
    for i := 0; i < n; i++ {
        if a[i] == zero {
            return [2]int{zero, one}
        }
    }
    return [2]int{one, zero}
}

File: SDE_sheet/Arrays/11_missing_and_repeating_number/solution.py
def findMissingRepeatingNumbers(a):
    n = len(a)
    xr = 0
    for i in range(n):
        xr ^= a[i]
        xr ^= (i + 1)
    number = xr & ~(xr - 1)
    zero, one = 0, 0
    for i in range(n):
        if (a[i] & number) != 0:
            one ^= a[i]
        else:
            zero ^= a[i]
    for i in range(1, n + 1):
        if (i & number) != 0:
            one ^= i
        else:
            zero ^= i
    for i in range(n):
        if a[i] == zero:
            return [zero, one]
    return [one, zero]

